### PR TITLE
Tracing: Fix incorrect indentations due to reoccurring spanIDs

### LINKED ISSUE #41573: Span are indented incorrectly in Tempo search split pane
**What happened**:
When using the Tempo search split pane and selecting different traces the indention and colors of the spans of the trace are sometimes rendered incorrectly. 

Things I've noticed about this behavior:
- The colors/indentation seems to be based on swapping between other traces in the same search results. It's as if some kind of state is not being completely cleared out between selections.
- If a trace is rendered incorrectly in the split pane closing the search pane so only the trace is visible always corrects the issue. I have not seen this issue in a full trace window.
- The example below shares details about how the first trace has issues, but I have seen this issue on different positions in the list.
- The example below shares details about how a single span indentation can occur but I have seen much more severely impacted traces as well as span color swapping.

This is the first trace in the search results (74b7a9794feaf48d) rendered correctly:

image

If I click the next 3 results in order and then come back it looks like this:

image

Note the incorrect indentation of the final span. If I then click the second trace in the search results and come back to the first in then it goes back to being rendered correctly. The full order of steps is:

1) Click first result (rendered correctly)
2) Click 2 -> 4 in order.
3) Click first result again (rendered incorrectly)
4) Click 2
5) Click first result again (rendered correctly)

**What you expected to happen**:
The colors and indention of the spans to be rendered consistently and correctly in the split pane view with Tempo search.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Grafana version: Grafana v8.2.3-40566 (3b8be57b4f)
- Data source type & version: Tempo
- OS Grafana is installed on: Hosted Grafana
- User OS & Browser: Ubuntu 18.04/Chrome
- Grafana plugins: 
- Others:

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR fixes incorrect indentation caused by the fact, that rows in the trace view didn't have unique ids. We have used only `spanId` as a unique identifier, but when old trace and new trace had span with the same `spanId`, the key for that row didn't change and react behaved unexpectedly -> wrong indentation & wrong ancestors. 

This PR fixes it by adding traceId into the key -> `${span.traceID}-${span.spanID}--${isDetail ? 'detail' : 'bar'}`. We expect that combination of `traceID` + `spanID` should be unique. 

Moreover, I have added a fix for the issue I discovered during working on this. We didn't clear detail state between the traces and therefore if we opened a detail view for span and then loaded a new trace with the same `spanId`, the detail view for that new `spanId` was opened as well.

I have also added tests for these fixes. 

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/41573

**Special notes for your reviewer**:
You can see here recording from main branch - it has 2 traces with the same spanID and in the second trace (trace2) you can see incorrect indentation.

https://user-images.githubusercontent.com/30407135/142589781-6c7a8ec3-711f-4901-a786-a33b4628413d.mov

Here is the recording from this (fixed) branch:

https://user-images.githubusercontent.com/30407135/142590006-f6561dc0-3f87-481b-93d4-77fab8374814.mov

To test this out, you can use these 2 sample traces that both have the same spanID. 

trace1:
```
{
  "batches": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "frontend"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "frontend-1"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "hIoiMMXij/Q=",
              "name": "frontend",
              "startTimeUnixNano": "1634296766948000000",
              "endTimeUnixNano": "1634296767232000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    },
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "querier"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "querier-1"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "MdmmKDpzjyE=",
              "parentSpanId": "hIoiMMXij/Q=",
              "name": "span from querier-1",
              "startTimeUnixNano": "1634296766990000000",
              "endTimeUnixNano": "1634296767167000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    },
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": {
              "stringValue": "querier"
            }
          },
          {
            "key": "host.name",
            "value": {
              "stringValue": "querier-2"
            }
          }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAACEiiIwxeKP9A==",
              "spanId": "mwJYmrj4CBs=",
              "parentSpanId": "hIoiMMXij/Q=",
              "name": "span from querier-2",
              "startTimeUnixNano": "1634296767133000000",
              "endTimeUnixNano": "1634296767147000000",
              "attributes": [],
              "status": {}
            }
          ]
        }
      ]
    }
  ]
}
```

trace2:
```
{
  "batches": [
    {
      "resource": {
        "attributes": [
          { "key": "service.name", "value": { "stringValue": "db" } },
          { "key": "job", "value": { "stringValue": "tns/db" } },
          { "key": "opencensus.exporterversion", "value": { "stringValue": "Jaeger-Go-2.22.1" } },
          { "key": "host.name", "value": { "stringValue": "63d16772b4a2" } },
          { "key": "ip", "value": { "stringValue": "0.0.0.0" } },
          { "key": "client-uuid", "value": { "stringValue": "39fb01637a579639" } }
        ]
      },
      "instrumentationLibrarySpans": [
        {
          "instrumentationLibrary": {},
          "spans": [
            {
              "traceId": "AAAAAAAAAABguiq7RPE+rg==",
              "spanId": "mwJYmrj4CBs=",
              "parentSpanId": "OY8PIaPbma4=",
              "name": "HTTP GET - root",
              "kind": "SPAN_KIND_SERVER",
              "startTimeUnixNano": "1627471657255809000",
              "endTimeUnixNano": "1627471657256268000",
              "attributes": [
                { "key": "http.status_code", "value": { "intValue": "200" } },
                { "key": "http.method", "value": { "stringValue": "GET" } },
                { "key": "http.url", "value": { "stringValue": "/" } },
                { "key": "component", "value": { "stringValue": "net/http" } }
              ],
              "status": {}
            }
          ]
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: public/app/features/explore/TraceView/TraceView.test.tsx [hunk 1]
```diff
@@ -143,6 +143,27 @@ describe('TraceView', () => {
...
+    userEvent.click(span);
+    //Process is in detail view
+    expect(screen.getByText(/Process/)).toBeInTheDocument();
+
+    rerender(
+      <Provider store={store}>
+        <TraceView exploreId={ExploreId.left} dataFrames={[frameNew]} splitOpenFn={() => {}} />
+      </Provider>
+    );
+    //Process is in detail view
```
Comment (connorlindsey) [inline public/app/features/explore/TraceView/TraceView.test.tsx:164]: nit: Process in not in view

**Review by connorlindsey — APPROVED** [review]
Tested on your branch and compared with ops. Looks great! Nice job finding the bug with the duplicate span id's! Makes sense in retrospect why the bug was happening but must've been hard to find

File: public/app/features/explore/TraceView/TraceView.test.tsx (hunk 1 above)
Comment (ivanahuckova) [reply public/app/features/explore/TraceView/TraceView.test.tsx:164]:
```suggestion
```

File: public/app/features/explore/TraceView/TraceView.tsx [hunk 2]
```diff
@@ -50,7 +56,16 @@ export function TraceView(props: Props) {
...
     detailStackTracesToggle,
+    clearDetailStates,
   } = useDetailState();
+
+  // Clear detail state when new trace arrives
+  useEffect(() => {
+    if (frame && prevFrame && !isEqual(prevFrame, frame)) {
+      clearDetailStates();
+    }
+  }, [frame, prevFrame, clearDetailStates]);
```
Comment (aocenas) [inline public/app/features/explore/TraceView/TraceView.tsx:67]: I think we could just have `frame` as dependency can't we? I mean if we don't need deep equal here, then the dependency array is checked whether it changes and basically giving you the `frame != prevFrame` check out of the box.

Comment (aocenas) [reply public/app/features/explore/TraceView/TraceView.tsx:67]: Also maybe it could make sense to put this effect inside `useDetailState`

**Comment by markoknez** [comment]
Not sure if this one is the culprit but reading the description I would guess that this PR produced now a bug where rows have wrong service name reported. TraceId/SpanId is not unique and it's used by zipkin to report the difference between client-send and server-receive (and opposite on the return path). You can find the mention of this in documentation - https://zipkin.io/pages/instrumenting.html. Under **When to set Span.timestamp and duration** it is mentioned 

> Zipkin merges spans together that share the same trace and span ID. The most common case of this is to merge a span reported by both the client (cs, cr) and the server (sr, ss). For example, the client starts a span, logging “cs” and propagates it via B3 headers, the server continues that span by logging “sr”.

**Comment by connorlindsey** [comment]
@markoknez If you're experiencing this bug, could you file a bug report in this repo with steps to reproduce the issue and we can take a look? Thanks

**Pull request merged**